File: tests/test_generate_endpoint.cpp
#include "cllm/model/executor.h"
#include "cllm/tokenizer/tokenizer_manager.h"
#include "cllm/scheduler/scheduler.h"
#include "cllm/endpoint/generate_endpoint.h"
#include "cllm/http/request.h"
#include "cllm/http/response.h"
#include "cllm/common/logger.h"
#include <nlohmann/json.hpp>
#include <iostream>
#include <string>
#include <memory>

using json = nlohmann::json;

int main() {
    try {
        // 设置日志
        cllm::Logger::instance().setLevel(cllm::LogLevel::INFO);
        
        // 创建模型配置
        cllm::ModelConfig modelConfig;
        modelConfig.vocabSize = 151936; // Qwen3 词表大小
        modelConfig.maxSequenceLength = 2048;
        modelConfig.defaultMaxTokens = 10;
        modelConfig.temperature = 0.7;
        modelConfig.topK = 50;
        modelConfig.topP = 0.9;
        
        // 创建模型执行器（使用空模型路径，启用占位权重模式）
        cllm::ModelExecutor modelExecutor("", "fp16", true, false);
        
        // 加载模型（实际会使用占位权重）
        modelExecutor.loadModel();
        
        // 初始化分词器（使用默认的 vocab.txt）
        // 注意：这里我们需要一个真实的分词器文件路径，或者模拟一个
        // 为了简单起见，我们直接使用一个不存在的路径，看看代码如何处理
        try {
            cllm::TokenizerManager tokenizerManager("/tmp", &modelExecutor);
            cllm::ITokenizer* tokenizer = tokenizerManager.getTokenizer();
            
            // 创建调度器
            cllm::Scheduler scheduler(&modelExecutor, 8, 2048);
            scheduler.start();
            
            // 创建生成端点
            cllm::GenerateEndpoint generateEndpoint(&scheduler, tokenizer);
            
            // 创建测试请求
            std::string requestBody = R"({
                "prompt": "hello",
                "max_tokens": 10,
                "temperature": 0.7,
                "top_k": 50,
                "top_p": 0.9
            })";
            
            cllm::HttpRequest request;
            request.setMethod(cllm::HttpMethod::POST);
            request.setPath("/generate");
            request.setBody(requestBody);
            request.setHeader("Content-Type", "application/json");
            
            // 处理请求
            std::unique_ptr<cllm::HttpResponse> response = generateEndpoint.handle(request);
            
            // 输出结果
            std::cout << "Status Code: " << static_cast<int>(response->getStatus()) << std::endl;
            std::cout << "Response Body: " << response->getBody() << std::endl;
            
            // 停止调度器
            scheduler.stop();
            
        } catch (const std::exception& e) {
            std::cerr << "Tokenizer initialization error: " << e.what() << std::endl;
            std::cerr << "Creating mock tokenizer instead..." << std::endl;
            
            // 创建一个非常简单的模拟分词器
            class MockTokenizer : public cllm::ITokenizer {
            public:
                size_t getVocabSize() const override { return 151936; }
                std::vector<int> encode(const std::string& text) const override {
                    // 简单的模拟编码
                    std::vector<int> tokens;
                    for (char c : text) {
                        tokens.push_back(static_cast<int>(c) + 100);
                    }
                    return tokens;
                }
                std::string decode(const std::vector<int>& tokens) const override {
                    // 简单的模拟解码
                    std::string text;
                    for (int token : tokens) {
                        text += static_cast<char>(token - 100);
                    }
                    return text;
                }
            };
            
            MockTokenizer mockTokenizer;
            
            // 创建调度器
            cllm::Scheduler scheduler(&modelExecutor, 8, 2048);
            scheduler.start();
            
            // 创建生成端点
            cllm::GenerateEndpoint generateEndpoint(&scheduler, &mockTokenizer);
            
            // 创建测试请求
            std::string requestBody = R"({
                "prompt": "hello",
                "max_tokens": 10,
                "temperature": 0.7,
                "top_k": 50,
                "top_p": 0.9
            })";
            
            cllm::HttpRequest request;
            request.setMethod(cllm::HttpMethod::POST);
            request.setPath("/generate");
            request.setBody(requestBody);
            request.setHeader("Content-Type", "application/json");
            
            // 处理请求
            std::unique_ptr<cllm::HttpResponse> response = generateEndpoint.handle(request);
            
            // 输出结果
            std::cout << "Status Code: " << static_cast<int>(response->getStatus()) << std::endl;
            std::cout << "Response Body: " << response->getBody() << std::endl;
            
            // 停止调度器
            scheduler.stop();
        }
        
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
}

File: tests/test_generate_endpoint.py
#!/usr/bin/env python3
"""
测试 /generate 接口的脚本
"""
import requests
import json
import sys

def test_generate_endpoint():
    """测试 /generate 接口"""
    print("="*60)
    print("测试 cLLM 服务器 /generate 接口")
    print("="*60)
    
    # 测试配置
    base_url = "http://localhost:8080"
    generate_url = f"{base_url}/generate"
    
    # 测试数据
    test_data = {
        "prompt": "hello",
        "max_tokens": 3,
        "temperature": 0.7
    }
    
    print(f"\n测试请求:")
    print(f"URL: {generate_url}")
    print(f"请求数据: {json.dumps(test_data, indent=2)}")
    
    try:
        # 发送请求
        response = requests.post(generate_url, json=test_data, timeout=10)
        
        print(f"\n响应结果:")
        print(f"状态码: {response.status_code}")
        print(f"响应头: {dict(response.headers)}")
        
        if response.status_code == 200:
            # 解析JSON响应
            try:
                response_data = response.json()
                print(f"响应体 (JSON): {json.dumps(response_data, indent=2, ensure_ascii=False)}")
                
                # 验证响应结构
                if "text" in response_data:
                    print(f"\n✅ 成功: 接口返回了生成的文本")
                    print(f"生成文本: {response_data['text']}")
                    return True
                else:
                    print(f"\n❌ 失败: 响应中缺少 'text' 字段")
                    return False
            except json.JSONDecodeError:
                print(f"\n响应体 (原始): {response.text}")
                print(f"❌ 失败: 响应不是有效的JSON格式")
                return False
        else:
            print(f"\n响应体: {response.text}")
            print(f"❌ 失败: 请求失败，状态码 {response.status_code}")
            return False
            
    except requests.exceptions.ConnectionError:
        print(f"\n❌ 失败: 无法连接到服务器 {base_url}")
        print("请确保服务器已启动并在端口 8080 上运行")
        return False
    except requests.exceptions.Timeout:
        print(f"\n❌ 失败: 请求超时")
        return False
    except Exception as e:
        print(f"\n❌ 失败: 发生未知错误: {e}")
        return False

def main():
    """主函数"""
    success = test_generate_endpoint()
    print("\n" + "="*60)
    if success:
        print("✅ 测试通过!")
        return 0
    else:
        print("❌ 测试失败!")
        return 1

if __name__ == "__main__":
    sys.exit(main())